### PR TITLE
TG-129: Task form validations and React lazy

### PR DESCRIPTION
**Pull request by mpueblas**
- Validations with Joi schema.
- React lazy working for a better perfomarnce.
- Absolute import for routes working.

### CODE REVIEW

**Review by martinalbrnz — COMMENTED** [review]
- validation messages are not working
- it's recommended to add a dropdown or a search to the parent project
- React lazy working fine

File: src/Components/Tasks/AddTask/index.jsx [hunk 1]
```diff
@@ -1,108 +1,180 @@
...
+    const previousState = selectedEmployees;
+    setSelectedEmployees([...previousState, id]);
+    setEmployees('');
+  };
+
+  const deleteFromSelectedEmployees = (id) => {
+    setSelectedEmployees(selectedEmployees.filter((emp) => emp !== id));
+  };
   const dispatch = useDispatch();
+  console.log(errors);
```
Comment (nicolascostanza) [inline src/Components/Tasks/AddTask/index.jsx:54]: console.log nop

File: src/Components/Tasks/index.js [hunk 2]
```diff
@@ -1,24 +1,56 @@
...
-import Table from '../Shared/Table';
-import Form from '../Shared/Form';
+import Table from 'Components/Shared/Table';
+import Form from 'Components/Shared/Form';
 import AddTask from './AddTask';
-import * as taskThunks from '../../redux/tasks/thunks';
-import * as employeesThunks from '../../redux/employees/thunks';
+import * as taskThunks from 'redux/tasks/thunks';
+import * as employeesThunks from 'redux/employees/thunks';
+import * as projectThunks from 'redux/projects/thunks';
```
Comment (nicolascostanza) [inline src/Components/Tasks/index.js:9]: excelente

File: src/Components/Tasks/index.js [hunk 3]
```diff
@@ -46,23 +78,40 @@ function Tasks() {
...
+        reset({
+          parentProject: response.data.parentProject?._id,
+          taskCreatorId: response.data.taskCreatorId,
+          taskName: response.data.taskName,
+          taskDescription: response.data.taskDescription,
+          assignedEmployee: response.data.assignedEmployee,
+          startDate: response.data.startDate,
+          status: response.data.status
+        });
+        console.log(response);
```
Comment (nicolascostanza) [inline src/Components/Tasks/index.js:100]: este console.log no va

File: src/Components/Tasks/AddTask/index.jsx [hunk 4]
```diff
@@ -1,108 +1,180 @@
...
+      .messages({ 'string.empty': 'This field is required' }),
+    taskName: Joi.string().min(1).max(50).required().messages({
+      'string.min': 'Name must contain 1 or more characters',
+      'string.max': 'Name must contain 50 or less characters',
+      'string.empty': 'This field is required'
+    }),
+    taskDescription: Joi.string().min(1).max(250).optional().messages({
+      'string.min': 'Name must contain 1 or more characters',
+      'string.max': 'Name must contain 250 or less characters'
+    }),
```
[the remark this answers is not in the document]
Comment (mpueblas) [reply src/Components/Tasks/AddTask/index.jsx:25]: it doesnt have a string.empty becouse is an optional field / min message fixed

File: src/Components/Tasks/AddTask/index.jsx [hunk 5]
```diff
@@ -1,108 +1,180 @@
...
+      ...data,
+      assignedEmployee: selectedEmployees
     });
-    setParentProject('');
-    setTaskName('');
-    setTaskDescription('');
-    setAssignedEmployee([]);
-    setStartDate('');
-    setStatus('');
+    console.log('data: ', data);
```
[the remark this answers is not in the document]
Comment (mpueblas) [reply src/Components/Tasks/AddTask/index.jsx:65]: fixed

File: src/Components/Tasks/AddTask/index.jsx (hunk 1 above)
Comment (mpueblas) [reply src/Components/Tasks/AddTask/index.jsx:54]: fixed

**Review by GinaSchP — DISMISSED** [review]

**Review by martinalbrnz — DISMISSED** [review]
LGTM

**Review by fernandocarbone1 — DISMISSED** [review]

**Review by karu-ams — CHANGES_REQUESTED** [review]
## TEST RESULTS ❗ 
Tested locally on _https://alfonso-trackgenix-nnp6l61zo-basp-m2022.vercel.app/_

**Test Results:** FAILED ⚠️ Bugs found

1. A load icon could be implemented so the "fetching..." is not seen by the user. This could be implemented too when the user entered in the edit form because the previuos data that has been charged is shown before the actual is load.
https://user-images.githubusercontent.com/101274200/175785973-8905944a-fd17-41c1-9b96-fbe86821c125.mp4

https://user-images.githubusercontent.com/101274200/175786108-08a80ad7-9c23-4b81-a844-9fcc25945a14.mp4

2. The validation of the inputs should be onblur to maintain consistency with the other entities.
chrome-capture-2022-5-25

3. As a suggestion: The form content in the modal could be centered, and status input could be the same size as the other ones.
image
image

**Review by karu-ams — APPROVED** [review]
## TEST RESULTS
Tested locally on _https://alfonso-trackgenix-1v60jzbjl-basp-m2022.vercel.app/_

Test Results: PASS ✅ READY TO MERGE

**Review by JuanBSantomero — DISMISSED** [review]
Everything looks great

**Review by JuanBSantomero — APPROVED** [review]

**Review by FefeTroanes — APPROVED** [review]
😎👉🏼

**Pull request merged**